raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1326,6 +1408,17 @@ private boolean handleFetchSnapshotResponse(
             // Finished fetching the snapshot.
             snapshot.freeze();
             state.setFetchingSnapshot(Optional.empty());
+
+            if (log.truncateFullyToLatestSnapshot()) {
+                updateFollowerHighWatermark(state, OptionalLong.of(log.highWatermark()));
+            } else {
+                logger.error(
+                    "Full log trunctation expected but didn't happend. Snapshot of {}, log end offset {}, last fetched {}",
```

Comment:
Good question. I am not sure. I think, since we pause regular fetch handling, if this happens then we have:

1. A bug in the Leader or the Follower:
  a. There is a bug in the leader if the returned snapshot id is less that the last fetched epoch and log end offset. 
  b. There is a bug in the follower if it sent a fetch offset and last fetched offset that doesn't match the local log.
2. The leader lost data at the end of the log.

What do you think? I am thinking we should throw an fatal exception like `IllegalStateException`.

Reply:
Updated the code to throw an exception instead of logging an error.